plutus-core/untyped-plutus-core/test/Transform/CaseOfCase/Test.hs

```diff
@@ -0,0 +1,121 @@
+      false = Constr () 1 []
+      alts = V.fromList [mkConstant @Integer () 1, mkConstant @Integer () 2]
+  pure $ Case () (mkIterApp ite [((), Var () b), ((), true), ((), false)]) alts
+
+{- | This should not simplify, because one of the branches of `ifThenElse` is not a `Constr`.
+Unless both branches are known constructors, the case-of-case transformation
+may increase the program size.
+-}
+caseOfCase2 :: Term Name PLC.DefaultUni PLC.DefaultFun ()
+caseOfCase2 = runQuote do
+  b <- freshName "b"
+  t <- freshName "t"
+  let ite = Force () (Builtin () PLC.IfThenElse)
+      true = Var () t
+      false = Constr () 1 []
+      alts = V.fromList [mkConstant @Integer () 1, mkConstant @Integer () 2]
+  pure $ Case () (mkIterApp ite [((), Var () b), ((), true), ((), false)]) alts
+
+{- | Similar to `caseOfCase1`, but the type of the @true@ and @false@ branches is
+@[Integer]@ rather than Bool (note that @Constr 0@ has two parameters, @x@ and @xs@).
+-}
+caseOfCase3 :: Term Name PLC.DefaultUni PLC.DefaultFun ()
+caseOfCase3 = runQuote do
+  b <- freshName "b"
+  x <- freshName "x"
+  xs <- freshName "xs"
+  f <- freshName "f"
+  let ite = Force () (Builtin () PLC.IfThenElse)
+      true = Constr () 0 [Var () x, Var () xs]
+      false = Constr () 1 []
+      altTrue = Var () f
+      altFalse = mkConstant @Integer () 2
+      alts = V.fromList [altTrue, altFalse]
+  pure $ Case () (mkIterApp ite [((), Var () b), ((), true), ((), false)]) alts
+
+{- |
+
+@
+  case (force ifThenElse) True True False of
+    True -> ()
+    False -> _|_
+@
+
+Evaluates to `()` because the first case alternative is selected.
+(The _|_ is not evaluated because case alternatives are evaluated lazily).
+
+After the `CaseOfCase` transformation the program should evaluate to `()` as well.
+
+@
+  force ((force ifThenElse) True (delay ()) (delay _|_))
+@
+
+-}
+caseOfCaseWithError :: TestTree
+caseOfCaseWithError =
+  testCase "Transformation doesn't evaluate error eagerly" do
+    let
+       originalTerm =
+        Case
+          ()
+          ( mkIterApp
+              (Force () (Builtin () PLC.IfThenElse))
+              [ ((), mkConstant @Bool () True)
+              , ((), Constr () 0 []) -- True
+              , ((), Constr () 1 []) -- False
+              ]
+          )
+          (V.fromList [mkConstant @() () (), Error ()])
+    simplifiedTerm <- runQuoteT $ simplifyTerm defaultSimplifyOpts def originalTerm
+    evaluateUplc simplifiedTerm @?= evaluateUplc originalTerm
+
+evaluateUplc
+  :: UPLC.Term Name DefaultUni DefaultFun ()
+  -> EvaluationResult (UPLC.Term Name DefaultUni DefaultFun ())
+evaluateUplc = fst <$> unsafeEvaluateCek noEmitter machineParameters
+ where
+  costModel :: CostModel CekMachineCosts BuiltinCostModel =
+    CostModel defaultCekMachineCosts defaultBuiltinCostModel
+  machineParameters
+    :: MachineParameters CekMachineCosts DefaultFun (CekValue DefaultUni DefaultFun ()) =
+      mkMachineParameters def costModel
```

Comment:
Mostly as a matter of taste 🤷🏼‍♂️ ,

```haskell
  costModel :: CostModel CekMachineCosts BuiltinCostModel =
  costModel = CostModel defaultCekMachineCosts defaultBuiltinCostModel
```
vs.

```haskell
  costModel :: CostModel CekMachineCosts BuiltinCostModel =
    CostModel defaultCekMachineCosts defaultBuiltinCostModel
```

but if I had to justify, I'd say that the latter 
1. occupies less horizontal space and 
2. doesn't repeat `costModel` twice (did you ever make a typo such that name in type sig doesn't match value-level name? "The type signature for `costModel` lacks an accompanying binding...")

Also, we're using two-char wide `::`   (instead of the 1-char wide `∷`) and it doesn't align vertically with the 1-char wide `=`. See how ugly it looks in the first example where `CostModel` is not on the same vertical line.

Reply:
> Also, we're using two-char wide `::`   (instead of the 1-char wide `∷`) and it doesn't align vertically with the 1-char wide `=`. See how ugly it looks in the first example where `CostModel` is not on the same vertical line.

I don't personally find it ugly, I find `=` at the end of a type ugly, but that's all a matter of taste of course.

But the point is that my version is a type _signature_ + a definition, while your version is a type _annotation_ + a definition. I.e. for example this:

```haskell
blah :: forall a. a -> a
blah = f where
    f :: forall b. b -> b
    f = \x -> x
```

type checks, while this:

```haskell
blah :: forall a. a -> a
blah = f where
    f :: forall b. b -> b =
        \x -> x
```

doesn't, giving

```
    • Couldn't match expected type: p0 -> p0
                  with actual type: forall b. b -> b
    • When checking that the pattern signature: forall b. b -> b
```

It's a different kind of thing and you're going against GHC by doing it this way, so don't be surprised if things don't work properly in more complex cases.

Also we use type signatures literally everywhere at the top level, why should non-top level be any different? I also try to use `StandaloneKindSignatures` in as many cases as possible, because the kind of nonsense GHC does at the type level is incomparable to what it does at the term level, so you pretty much have to be very pedantic at the type level, otherwise you're gonna suffer a lot (speaking from personal experience).

All in all, you're going to have an overall more pleasant experience by doing things the way GHC expects you to.

But I don't mind it being written like you did, I'm just not going to comply with any guideline saying I need to do it this way too, should we have such a guideline.